File: JULY-2025/08-07-2025/p1751.ex
defmodule Solution do
  @spec max_value(events :: [[integer]], k :: integer) :: integer
  def max_value(events, k) do
    events = events |> Enum.sort() |> Enum.with_index()

    Task.async(fn ->
      solve(events, k, -1)
    end)
    |> Task.await()
  end

  # If there's no events to attend, return 0 as the total value.
  defp solve([], _k, _prev_finish), do: 0
  
  # If my quota is used up, return 0 as the total value.
  defp solve(_events, 0, _prev_finish), do: 0

  # If the current event starts before the previous event ends,
  # skip this event.
  defp solve([{[start, _finish, _value], _i} | rest_events], k, prev_finish)
       when start <= prev_finish,
       do: solve(rest_events, k, prev_finish)

  # Otherwise, try both attending and skipping the current event,
  # see which case I can get a greater total value.
  defp solve([{[_start, finish, value], i} | rest_events], k, prev_finish) do
    memoized({i, k}, fn ->
      attend = value + solve(rest_events, k - 1, finish)
      skip = solve(rest_events, k, prev_finish)
      max(attend, skip)
    end)
  end

  defp memoized(key, fun) do
    with nil <- Process.get(key) do
      fun.() |> tap(&Process.put(key, &1))
    end
  end
end

File: JULY-2025/08-07-2025/p1751.ts
function maxValue(events: number[][], k: number): number {
    events.sort((a, b) => a[0] - b[0]);
    const n = events.length;
    const starts = events.map(e => e[0]);
    const next_idx = new Array(n).fill(0);
    for (let i = 0; i < n; i++){
        let l = i + 1;
        let r = n;
        while (l < r){
            let m = Math.floor((l + r) / 2);
            if (starts[m] > events[i][1])
                r = m;
            else 
                l = m + 1;
        }
        next_idx[i] = l;
    }
    const dp = Array.from({length : 2}, () => new Array(n + 1).fill(0));
    for (let j = 1; j <= k; j++){
        const cur = j % 2;
        const prev = (j - 1) % 2;
        for (let i = n - 1; i >= 0; i--){
            const attend = events[i][2] + dp[prev][next_idx[i]];
            const skip = dp[cur][i + 1];
            dp[cur][i] = Math.max(attend, skip);
        }
    }
    return dp[k % 2][0];
};